### PR TITLE
@W-10225659: Add warning message for jdk bug JDK-8209178

### PR DESCRIPTION
**Pull request by arjun-mehrotra**
1. Add a warning message when the `java.version` is not one of the versions where the bug is fixed. 
2. Added a version class with the capability to parse different `java.version` and perform symntic version comparison.

### CODE REVIEW

File: src/main/java/com/sforce/ws/types/Version.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.sforce.ws.types;
```
Comment (mgrumbach) [inline src/main/java/com/sforce/ws/types/Version.java:26]: This class is not related to the other classes in this package. Please move it to the ws package, alongside ConnectorConfig.

File: src/main/java/com/sforce/ws/types/Version.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Version implements Comparable<Version> {
```
Comment (mgrumbach) [inline src/main/java/com/sforce/ws/types/Version.java:32]: It would be good to rename this class to "JavaVersion", to avoid confusion with API version.

Comment (mgrumbach) [inline src/main/java/com/sforce/ws/types/Version.java:32]: Please add a JavaDoc to the class.

File: src/main/java/com/sforce/ws/ConnectorConfig.java [hunk 3]
```diff
@@ -49,6 +53,9 @@
  */
 public class ConnectorConfig {
 
+    private final static String JAVA_VERSION_PROPERTY = "java.version";
+    private final static String JDK_UPDATE_MESSAGE = "Please update the client to a JDK version that includes a fix for JDK-8209178. See https://trailblazer.salesforce.com/issues_view?id=a1p4V000002JfxSQAS for more details.";
+
```
Comment (mgrumbach) [inline src/main/java/com/sforce/ws/ConnectorConfig.java:58]: Most of the code added to this class is not related to configuring the connection, so I think it should be moved to another class. One option is to move it to the Version class, which I think would be fine. But if you prefer you could create another class (something like VersionUtil), which would also be fine). Mainly, I want to avoid adding clutter to ConnectorConfig.

File: src/main/java/com/sforce/ws/ConnectorConfig.java [hunk 4]
```diff
@@ -161,9 +168,23 @@ public void close() throws IOException {
...
 	  private SSLContext sslContext;
-
+    private final static Version VERSION_14_0_0_0 =
+            Version.Builder.newBuilder().setFeature(14).build();
+    private final static Version VERSION_13_0_2 =
+            Version.Builder.newBuilder().setFeature(13).setInterim(0).setUpdate(2).build();
+    private final static Version VERSION_11_0_6 =
+            Version.Builder.newBuilder().setFeature(11).setInterim(0).setUpdate(6).build();
+    private final static Version VERSION_1_8_0_0_321 =
+            Version.Builder.newBuilder().setFeature(1).setInterim(8).setUpdate(0).setPatch(0).setBuild(321).build();
```
Comment (mgrumbach) [inline src/main/java/com/sforce/ws/ConnectorConfig.java:178]: All of the above constants can be moved to another class.

File: src/main/java/com/sforce/ws/ConnectorConfig.java [hunk 5]
```diff
@@ -161,9 +168,23 @@ public void close() throws IOException {
...
+    private static boolean javaVersionHasBug = doesVersionHasABug(System.getProperty(JAVA_VERSION_PROPERTY));
```
Comment (mgrumbach) [inline src/main/java/com/sforce/ws/ConnectorConfig.java:179]: Another option is to just call doesVersionHasABug() twice. Caching it here does not save very much.

File: src/main/java/com/sforce/ws/ConnectorConfig.java [hunk 6]
```diff
@@ -549,4 +573,27 @@ public void setCaptureHtmlExceptionResponse(boolean captureHtmlResponse) {
...
+
+    /**
+     * Check if the provided java version has been patched for the bug JDK-8209178. Refer
+     * <a href="https://bugs.openjdk.org/browse/JDK-8209178">JDK-8209178</a> to get the list of java versions that have
+     * fixed the bug.
+     *
+     * @param runtimeVersion java version
+     * @return true, if the java version has been patched for the bug, false otherwise
+     */
+    public static boolean doesVersionHasABug(String runtimeVersion) {
```
Comment (mgrumbach) [inline src/main/java/com/sforce/ws/ConnectorConfig.java:585]: The name of this method sounds odd. Maybe just "versionHasABug()) would be better.

File: src/main/java/com/sforce/ws/ConnectorConfig.java (hunk 5 above)
Comment (mgrumbach) [inline src/main/java/com/sforce/ws/ConnectorConfig.java:179]: I think it would be good to overload this method:
```
boolean doesVersionHasABug() {
    return doesVersionHasABug(System.getProperty(JAVA_VERSION_PROPERTY));
}
```

**Comment by arjun-mehrotra** [comment]
@mgrumbach Thank you for your feedback. I agree with the suggestions and have updated my PR.

**Review by mgrumbach — APPROVED** [review]
